---
title: "Analysis workflow for brain development data"
output: html_document
---

# Load packages

```{r, message = FALSE, warning = FALSE}
library(ggplot2)
library(tidyr)
library(dplyr)
library(ComplexHeatmap)
library(circlize)
library(dendsort)
library(dendextend)
```

# Data description

BrainSpan: Atlas of the developing human brain.
This dataset contains RNA-Seq RPKM (reads per kilobase per million.
See the whitepaper at <https://www.brainspan.org/>) values averaged to genes.

- `expression_matrix.csv` - The rows are genes and the columns samples;
  the first column is the row number.
- `rows_metadata.csv` - The genes are listed in the same order as the rows
  in `expression_matrix.csv`.
- `columns_metadata.csv` - The samples are listed in the same order as
  the columns in `expression_matrix.csv`.

# Download data

Data manually downloaded from
<https://www.brainspan.org/api/v2/well_known_file_download/267666525>.
Expression and gene list data from OMIM.

```{r}
rm(list = ls())
# Expression data
expression <- read.csv("./datadownloaded/expression_matrix.csv", header = FALSE)

# Read meta data
row <- read.csv("./datadownloaded/rows_metadata.csv")
column <- read.csv("./datadownloaded/columns_metadata.csv")

# Combine
dat <- data.frame(row, expression[, -1])

# Select epilepsy genes
gene <- read.csv(file = "genelist_all_updated_group2.csv")
gene <- gene[, -1]
colnames(gene) <- c("gene_symbol", "Group")

dat_epi <- left_join(gene, dat)
dat_epi <- na.omit(dat_epi)

dat_epi_l <- dat_epi %>%
  gather("column_num", "value", 7:ncol(dat_epi))

dat_epi_l$column_num <- gsub("V", "", dat_epi_l$column_num)
dat_epi_l$column_num <- as.numeric(dat_epi_l$column_num) - 1

# Add age info
dat_f <- left_join(dat_epi_l, column)

write.csv(dat_f, file = "summary.csv")
```

# Group into 5 groups

```{r}
rm(list = ls())
data <- read.csv("summary.csv")

Prenatal <- c("8 pcw", "9 pcw", "12 pcw", "13 pcw", "16 pcw", "17 pcw", "19 pcw", "21 pcw", "24 pcw", "25 pcw", "26 pcw", "35 pcw", "37 pcw")
Infancy <- c("4 mos", "10 mos", "1 yrs")
Childhood <- c("2 yrs", "3 yrs", "4 yrs", "1 yrs", "8 yrs", "11 yrs")
Adolescence <- c("13 yrs", "15 yrs", "18 yrs", "19 yrs")
Adulthood <- c("21 yrs", "23 yrs", "30 yrs", "36 yrs", "37 yrs", "40 yrs")

for (i in 1:nrow(data)) {
  if (data$age[i] %in% Prenatal) {
    data$Stage[i] <- "Prenatal"
  } else if (data$age[i] %in% Infancy) {
    data$Stage[i] <- "Infancy"
  } else if (data$age[i] %in% Childhood) {
    data$Stage[i] <- "Childhood"
  } else if (data$age[i] %in% Adolescence) {
    data$Stage[i] <- "Adolescence"
  } else if (data$age[i] %in% Adulthood) {
    data$Stage[i] <- "Adulthood"
  } else {
    data$Stage[i] <- "TBD"
  }
}

write.table(data, file = "summary_addBiggroup.txt", sep = ",")
```

# Area

```{r, warnings=FALSE}
options(dplyr.summarise.inform = FALSE)

rm(list = ls())

data <- read.table(file = "summary_addBiggroup.txt", sep = ",")
data <- data[, -1]

area <- names(table(data$structure_acronym))
write.csv(area, file = "arealist.csv")

for (j in 1:length(area)) {
  # select area and get mean value
  data_ave <- na.omit(data) %>%
    filter(structure_acronym == area[j]) %>%
    group_by(Group, gene_symbol, Stage) %>%
    summarise(ave = ave(value))

  data_ave <- unique(data_ave)

  write.csv(data_ave, file = paste("./dataProcessed/", area[j], "_Biggroup.csv", sep = ""))
}
```

# Heatmap

```{r}
set.seed(2023)
rm(list = ls())

area <- read.csv(file = "arealist.csv")
area <- area[, -1]
area <- area[-c(3, 5, 7, 11, 13, 16, 17, 19, 23, 24)] # Remove tissues with sparse data
length(area)

for (i in 1:length(area)) {
  path <- paste("./dataProcessed/", area[i], "_Biggroup.csv", sep = "")
  print(path)
  dat <- read.csv(path)
  dat <- dat[, -1]
  dat_w <- spread(dat, key = Stage, value = ave)
  dat_w[dat_w == 0] <- NA
  dat_w <- na.omit(dat_w)
  area0 <- area[i]

  Group <- dat_w[, 1]
  gene <- dat_w[, 2]
  Stage <- unique(dat$Stage)
  Stage <- c("Prenatal", "Infancy", "Childhood", "Adolescence", "Adulthood")

  heat <- as.matrix(dat_w[, -c(1:2)])
  heat <- log2(heat)

  # Create annotation: gene labels
  genelabels <- rowAnnotation(
    Genes = anno_mark(
      at = seq(1, nrow(heat), 1),
      labels = gene[seq(1, nrow(heat), 1)],
      labels_gp = gpar(fontsize = 2.5),
      padding = 0.5
    ),
    width = unit(1, "cm") +

      max_text_width(
        rownames(heat)[seq(1, nrow(heat), 5)],
        gp = gpar(fontsize = 4.5)
      )
  )

  # Set color scheme and choose breaks
  myCol <- colorRamp2(c(min(heat), median(heat), max(heat)), c("#008837", "white", "#7b3294"))

  row_dend <- dendsort(hclust(dist(heat)), isReverse = FALSE, type = "average")

  # creat heatmap object
  ht_list <- Heatmap(heat,
    col = myCol,
    name = "Expression",
    row_gap = unit(1, "mm"),
    column_gap = unit(1, "mm"),
    row_title = NULL,
    cluster_rows = FALSE,
    column_title = NULL,
    # cluster_columns = col_dend,
    cluster_columns = FALSE,
    show_column_dend = TRUE,
    show_column_names = TRUE, 
    column_names_gp = gpar(fontsize = 10),
    column_names_rot = 30,
    column_order = Stage,
    clustering_method_rows = "ward.D2",
    show_heatmap_legend = TRUE,
    heatmap_legend_param = list(
      title = "Expression",
      at = c(min(heat), median(heat), max(heat)),
      title_gp = gpar(fontsize = 10),
      labels = c("low", "median", "high"),
      labels_gp = gpar(fontsize = 8),
      legend_height = unit(2, "cm"),
      legend_width = unit(0.5, "cm")
    )
  ) +
    Heatmap(Group,
      col = c("#ffb60f", "#c13832", "#673bb8"),
      name = "Group",
      width = unit(2, "mm"),
      show_column_names = FALSE,
      show_heatmap_legend = TRUE, 
      heatmap_legend_param = list(
        title = "Group",
        title_gp = gpar(fontsize = 10),
        labels_gp = gpar(fontsize = 7),
        legend_height = unit(2, "cm"),
        legend_width = unit(0.5, "cm")
      )
    )

  png(filename = paste("./plot/", area0, "_Biggroup.png", sep = ""), width = 3.4, height = 3.4, units = "in", res = 300)
  draw(ht_list, main_heatmap = "Expression", heatmap_legend_side = "left", annotation_legend_side = "left")

  dev.off()
}
```

# Calculate mean values - hippocampus

```{r}
data <- read.csv("./dataProcessed/HIP_Biggroup.csv", sep = ",")
data <- data[, -1]

ave <- data %>%
  group_by(Group, Stage) %>%
  summarise(
    group_mean = mean(ave),
    total = n(),
    sd = sd(ave),
    sem = sd / sqrt(total)
  )
ave$Stage <- factor(
  ave$Stage,
  levels = c("Prenatal", "Infancy", "Childhood", "Adolescence", "Adulthood")
)

f <- ggplot(data = ave, aes(x = Stage, y = group_mean, group = Group)) +
  geom_point(aes(colour = Group), size = 3) +
  geom_smooth(aes(colour = Group), size = 1.5) +
  scale_color_manual(values = c("#ffb60f", "#c13832", "#673bb8")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank(),
    axis.line = element_line(colour = "black", size = 1),
    axis.ticks = element_line(colour = "black", size = 1.2),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.text = element_text(size = 15),
    axis.title = element_text(size = 25),
    strip.text.x = element_text(size = 25)
  ) + # Panel label font size
  ylab("Expression")

print(f)
ggsave(file = "./plot/mean_HIP.pdf", width = 5, height = 3)
```

# Mean calculation

```{r}
set.seed(2023)
rm(list = ls())

area <- read.csv(file = "arealist.csv")
area <- area[, -1]
area <- area[-c(3, 5, 7, 11, 13, 16, 17, 19, 23, 24)] # Remove tissues with sparse data
length(area)


for (i in 1:length(area)) {
  area0 <- area[i]

  path <- paste("./dataProcessed/", area[i], "_Biggroup.csv", sep = "")
  print(path)
  data <- read.csv(path)
  data <- data[, -1]

  ave <- data %>%
    group_by(Group, Stage) %>%
    summarise(group_mean = mean(ave), total = n(), sd = sd(ave), sem = sd / sqrt(total))
  ave$Stage <- factor(
    ave$Stage,
    levels = c("Prenatal", "Infancy", "Childhood", "Adolescence", "Adulthood")
  )

  f <- ggplot(data = ave, aes(x = Stage, y = group_mean, group = Group)) +
    geom_smooth(aes(colour = Group), size = 1.5) +
    scale_color_manual(values = c("#ffb60f", "#c13832", "#673bb8")) +
    theme_bw() +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      axis.line = element_line(colour = "black", size = 1),
      axis.ticks = element_line(colour = "black", size = 1.2),
      axis.title.x = element_blank(),
      axis.text.x = element_blank(),
      legend.position = "none",
      axis.text = element_text(size = 15),
      axis.title = element_text(size = 15),
      strip.text.x = element_text(size = 10)
    ) + # Panel label font size
    ylab("Expression")

  ggsave(filename = paste("./plot/", area0, "_mean.png", sep = ""), width = 5, height = 3)
}
```

# Statistical analysis for all brain regions

```{r}
data <- read.table(file = "summary_addBiggroup.txt", sep = ",")
data <- data[, -1]

stage <- c("Adulthood", "Adolescence", "Childhood", "Infancy", "Prenatal")
area <- read.csv(file = "arealist.csv")
area <- area[, -1]
region <- area[-c(3, 5, 7, 11, 13, 16, 17, 19, 23, 24)] # Remove tissues with sparse data
length(region)

for (i in 1:nrow(data)) {
  s <- stage[i]
  data1 <- data %>%
    filter(Stage == s)
  for (j in 1:nrow(data1)) {
    r <- region[j]
    data2 <- data1 %>%
      filter(structure_acronym == r)
    t <- pairwise.wilcox.test(data2$value, data2$Group, p.adjust.method = "BH")
    p0 <- t$p.value
    write.csv(p0, file = paste("./pvalue/", s, "_", r, ".csv", sep = ""))
  }
}
```

# Plot p-values from all brain regions

```{r}
rm(list = ls())
list <- list.files("./pvalue/", pattern = ".csv")

pvalue <- data.frame()
for (i in 1:length(list)) {
  l <- list[i]
  a <- gsub(".csv", "", l)
  b <- strsplit(a, "_")
  region <- b[[1]][1]
  stage <- b[[1]][2]
  r_s <- a
  path <- paste("./pvalue/", l, sep = "")
  d0 <- read.csv(path)
  DEEG_SRG0 <- d0[2, 3]
  DEEG_CEG0 <- d0[1, 2]
  CEG_SRG0 <- d0[2, 2]
  all <- cbind(region, stage, r_s, DEEG_SRG0, DEEG_CEG0, CEG_SRG0)
  pvalue <- rbind(pvalue, all)
}

write.csv(pvalue, file = "pvalue_sum.csv")

# Plot
data <- read.csv(file = "pvalue_sum.csv")
data <- data[, -1]

data_l <- data %>%
  gather(4:6, key = term, value = pvalue)

l1 <- 0.05 / 80
l2 <- 0.01 / 80
l3 <- 0.001 / 80
l4 <- 0.0001 / 80

for (i in 1:nrow(data_l)) {
  if (data_l$pvalue[i] > l1) {
    data_l$color[i] <- "ns"
  } else if (data_l$pvalue[i] < l1 && data_l$pvalue[i] >= l2) {
    data_l$color[i] <- "one"
  } else if (data_l$pvalue[i] < l2 && data_l$pvalue[i] >= l3) {
    data_l$color[i] <- "two"
  } else if (data_l$pvalue[i] < l3 && data_l$pvalue[i] >= l4) {
    data_l$color[i] <- "three"
  } else if (data_l$pvalue[i] < l4) {
    data_l$color[i] <- "four"
  }
}

data_l$color <- factor(data_l$color, levels = c("ns", "one", "two", "three", "four"))

f <- ggplot(data = data_l, aes(y = r_s, x = term, fill = color)) +
  geom_tile(color = "black") +
  scale_fill_manual(values = c(
    ns = "#FFFFFF", one = "#ffffd4",
    two = "#fed98e",
    three = "#fe9929",
    four = "#cc4c02"
  )) +
  scale_x_discrete(labels = c("CEG vs. SRG", "DEEG vs. CEG", "DEEG vs. SRG")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    axis.line = element_line(colour = "black", size = 1),
    axis.ticks = element_line(colour = "black", size = 1.2),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 10)
  ) + 
  ylab("Region-Stage")

print(f)
ggsave(f, file = "./plot/pvalue.png", width = 5, height = 10)
```

# Plot Hippocampus

```{r}
data <- read.csv("./dataProcessed/HIP_Biggroup.csv", sep = ",")

data <- data[, -1]
data$Stage <- factor(
  data$Stage,
  levels = c("Prenatal", "Infancy", "Childhood", "Adolescence", "Adulthood")
)
data$Group <- factor(data$Group, levels = c("DEEG", "CEG", "SRG"))

f <- ggplot(data = data, aes(x = Stage, y = ave, fill = Group)) +
  geom_boxplot(outlier.size = 0.5) +
  scale_fill_manual(values = c("#c13832", "#ffb60f", "#673bb8")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    axis.line = element_line(colour = "black", size = 1),
    axis.ticks = element_line(colour = "black", size = 1.2),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 8),
    strip.text.x = element_text(size = 25)
  ) + # Panel label font size
  ylab("Expression")

print(f)
ggsave(file = "./plot/HIP_box.png", width = 5, height = 3)
```

# Plot ten most studied genes

```{r}
rm(list = ls())

area <- read.csv(file = "arealist.csv")
area <- area[, -1]
region <- area[-c(3, 5, 7, 11, 13, 16, 17, 19, 23, 24)] # Remove tissues with sparse data

data <- read.table(file = "summary_addBiggroup.txt", sep = ",")

data <- data[, -1] %>% filter(structure_acronym %in% region)

topgene <- c(
  "CHRNA4", "CSTB", "DEPDC5", "EPM2A", "KCNQ2",
  "KCNT1", "SCN1A", "SCN2A", "SCN8A", "KCNA2"
)

for (j in 1:length(topgene)) {
  g <- topgene[j]
  dat <- data %>%
    filter(gene_symbol == g)

  dat$Stage <- factor(
    dat$Stage,
    levels = c("Prenatal", "Infancy", "Childhood", "Adolescence", "Adulthood")
  )

  f <- ggplot(dat, aes(x = Stage, y = value)) +
    geom_boxplot() +
    facet_wrap(. ~ structure_acronym, nrow = 4) +
    theme_bw() +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      axis.line = element_line(colour = "black", size = 0.5),
      axis.ticks = element_line(colour = "black", size = 0.5),
      axis.title.x = element_blank(),
      legend.position = "none",
      axis.text.y = element_text(size = 8),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
    ) + 
    ylab("Expression level") +
    ggtitle(g)

  path <- paste("./plot/", g, "_box.png", sep = "")
  ggsave(f, file = path, width = 6.8, height = 6.8)
}
```
