---
title: "Analysis workflow for GTEx data"
output: html_document
---

# Load packages

```{r, message = FALSE, warning = FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
library(xlsx)
library(RColorBrewer)
library(ComplexHeatmap)
library(circlize)
library(dendsort)
library(dendextend)
```

# Download GTEx data

Download Genotype-Tissue Expression (GTEx) data from the
[GTEx Portal](https://www.gtexportal.org/home/datasets).
Version: GTEx analysis V8.

Normalized quantification (TPM) across transcripts for each sample using
GENCODEv26. Quantifications were carried out using flair quantify.

# Annotations

```{r, echo = FALSE, warning = FALSE}
# Sample attributes
url1 <- "https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDD.xlsx"
dest1 <- "SampleAttributesDS.xlsx"
download.file(url1, dest1)

# Subject phenotypes
url2 <- "https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDD.xlsx"
dest2 <- "SubjectPhenotypesDS.xlsx"
download.file(url2, dest2)
```

# RNA-seq data (gene TPMs by tissue)

```{r}
list <- read.csv("list_tissue.csv")

for (i in 1:nrow(list)) {
  a <- list[i, 1]
  link <- paste("https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/gene_tpm/", a, sep = "")
  url <- link
  desk <- paste("./download_GTEX/", a, sep = "")
  download.file(url, desk)
}
```

# Get tissue names

```{r}
# tissue
list <- read.csv("list_tissue.csv")
name <- as.character(list$Name)

tissue <- str_extract_all(name, "(?<=v8_).+(?=.gct)")
tissue <- as.character(tissue)
write.csv(tissue, file = "tissuenames.csv")
```

# Get expression data for genes

```{r}
rm(list = ls())
list_exp <- list.files(path = "./downlaad_GTEX/", pattern = ".gct.gz")
tissue <- read.csv("tissuenames.csv")

# Mean expression
for (i in 1:nrow(tissue)) {
  a <- tissue[, 2][i]
  b <- paste("./download_GTEX/gene_tpm_2017-06-05_v8_", a, ".gct.gz", sep = "")
  dat.gct <- read.delim(file = b, skip = 2)
  c <- rowMeans(dat.gct[, -c(1:3)])
  d <- dat.gct[, c(3)]
  data <- cbind(d, c)
  colnames(data) <- c("Gene", "Expression_mean")
  write.table(data, file = paste("./Data_processed/meanExp_", a, ".txt", sep = ""))
}
```

# Subset all epilepsy genes

```{r}
# Read genes
rm(list = ls())
tissue <- read.csv("tissuenames.csv")

gene <- read.csv(file = "genelist_all_updated_group2.csv")
gene <- gene[, -1]
names(gene) <- c("Gene", "Group")

for (i in 1:nrow(tissue)) {
  a <- tissue[, 2][i]
  b <- paste("./Data_processed/meanExp_", a, ".txt", sep = "")
  c <- read.table(b, sep = " ")
  d <- left_join(gene, c)
  write.table(d, file = paste("./epilepsygenes/epilepsy_meanExp_", a, ".txt", sep = ""))
}

# Combine
dat <- data.frame(id = c(1:247))
for (i in 1:nrow(tissue)) {
  a <- tissue[, 2][i]
  b2 <- paste("./epilepsygenes/epilepsy_meanExp_", a, ".txt", sep = "")
  c <- read.table(b2, sep = " ")
  dat <- cbind(dat, c[, 3])
}

dat <- data.frame(gene, dat[, -1])
colnames(dat) <- c("Gene", "Group", as.character(tissue[, 2]))

# overlap = read.table(file = "genes_overlap.txt", sep = ",")
#
# dat = dat[!dat$Gene %in% overlap$Gene,] # remove overlapped genes

write.csv(dat, file = "./epilepsygenes/epilepsy_meanExp_sum.csv")
```

# Heatmap

```{r}
set.seed(2022)
rm(list = ls())

data <- read.csv(file = "./epilepsygenes/epilepsy_meanExp_sum.csv")
data <- data[, -c(1, 25:26)] # remove unnecessary columns

data[data == 0] <- NA
data <- na.omit(data)

meta <- data$Group

heat <- as.matrix(data[, -c(1, 2)])
heat <- log2(heat)
range(heat)

# Set color scheme and choose breaks
myCol <- colorRamp2(c(min(heat), median(heat), max(heat)), c("#0570b0", "white", "#d7301f"))

# Create annotation: gene labels
genelabels <- rowAnnotation(
  Genes = anno_mark(
    at = seq(1, nrow(heat), 1),
    labels = data$Gene[seq(1, nrow(heat), 1)],
    labels_gp = gpar(fontsize = 2.3),
    padding = 0.5
  ),
  width = unit(1, "cm")
)

col_dend <- dendsort(hclust(dist(t(heat)), method = "ward.D2"), isReverse = FALSE, type = "average")

# Create heatmap object
ht_list <- Heatmap(
  heat,
  col = myCol,
  name = "Expression",
  row_title = NULL,
  cluster_rows = TRUE,
  show_row_names = TRUE,
  column_title = NULL,
  cluster_columns = col_dend,
  show_column_dend = TRUE,
  show_column_names = TRUE,
  column_names_gp = gpar(fontsize = 6),
  show_parent_dend_line = TRUE,
  clustering_method_rows = "ward.D2",
  show_heatmap_legend = TRUE,
  heatmap_legend_param = list(
    title = "Expression",
    at = c(min(heat), median(heat), max(heat)),
    title_gp = gpar(fontsize = 9),
    labels = c("low", "median", "high"),
    labels_gp = gpar(fontsize = 8),
    legend_height = unit(2, "cm"),
    legend_width = unit(0.5, "cm")
  )
) +
  Heatmap(
    meta,
    col = c("#ffb60f", "#c13832", "#673bb8"),
    name = "Group",
    width = unit(4, "mm"),
    show_column_names = FALSE,
    show_heatmap_legend = TRUE,
    heatmap_legend_param = list(
      title = "Group",
      title_gp = gpar(fontsize = 9),
      labels_gp = gpar(fontsize = 8),
      legend_height = unit(2, "cm"),
      legend_width = unit(0.5, "cm")
    )
  )

png(file = "./plot/clusters_2_name_cluster.png", width = 6.8, height = 10, units = "in", res = 300)

draw(
  ht_list + genelabels,
  main_heatmap = "Expression",
  heatmap_legend_side = "left",
  annotation_legend_side = "left"
)
dev.off()
```

# Calculate mean values

```{r}
rm(list = ls())

data <- read.csv(file = "./epilepsygenes/epilepsy_meanExp_sum.csv")
data <- na.omit(data)
data <- data[, -c(1, 25:26)]

data_l <- data %>% gather(3:ncol(data), key = "tissue", value = "expression")

tissue <- unique(data_l$tissue)

a <- c(
  "brain_cerebellar_hemisphere", "brain_cerebellum", "brain_cortex",
  "brain_frontal_cortex_ba9", "brain_anterior_cingulate_cortex_ba24",
  "brain_caudate_basal_ganglia", "brain_putamen_basal_ganglia", "brain_nucleus_accumbens_basal_ganglia", "brain_amygdala", "brain_hippocampus",
  "brain_hypothalamus", "brain_spinal_cord_cervical_c.1",
  "brain_substantia_nigra", "pituitary", "testis",
  "esophagus_gastroesophageal_junction", "esophagus_muscularis",
  "colon_transverse", "small_intestine_terminal_ileum", "colon_sigmoid",
  "stomach", "artery_aorta", "artery_tibial", "artery_coronary",
  "adipose_subcutaneous", "adipose_visceral_omentum", "breast_mammary_tissue",
  "nerve_tibial", "cervix_ectocervix", "cervix_endocervix", "uterus", "bladder",
  "vagina", "prostate", "fallopian_tube", "ovary", "lung", "thyroid",
  "skin_not_sun_exposed_suprapubic", "skin_sun_exposed_lower_leg",
  "esophagus_mucosa", "minor_salivary_gland", "adrenal_gland", "kidney_cortex",
  "kidney_medulla", "spleen", "heart_atrial_appendage", "heart_left_ventricle",
  "muscle_skeletal", "liver", "pancreas", "whole_blood"
)

ave <- data_l %>%
  filter(tissue %in% a) %>%
  group_by(Group, tissue) %>%
  summarise(
    group_mean = mean(expression),
    total = n(),
    sd = sd(expression),
    sem = sd / sqrt(total)
  )

ave$tissue <- factor(ave$tissue, levels = a)

f <- ggplot(data = ave, aes(x = tissue, y = group_mean, group = Group)) +
  geom_bar(stat = "identity", position = position_dodge(), aes(fill = Group)) +
  scale_color_manual(values = c("#ffb60f", "#c13832", "#673bb8")) +
  scale_fill_manual(values = c("#ffb60f", "#c13832", "#673bb8")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    axis.line = element_line(colour = "black", size = 0.5),
    axis.ticks = element_line(colour = "black", size = 0.5),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 9)
  ) 
  ylab("Expression level")

print(f)
ggsave(file = "./plot/expression_mean.png", width = 6, height = 2.5)
```

# Plot ten most studied genes

# Get expression data for 10 genes 
```{r}
rm(list = ls())

list_exp <- list.files(path = "./downlaad_GTEX/", pattern = ".gct.gz")
tissue <- read.csv("tissuenames.csv")

topgene <- c(
  "CHRNA4", "CSTB", "DEPDC5", "EPM2A", "KCNQ2",
  "KCNT1", "SCN1A", "SCN2A", "SCN8A", "KCNA2"
)

# Gene data from all tissues
topgene_exp <- data.frame()
for (i in 1:nrow(tissue)) {
  a <- tissue[, 2][i]
  b <- paste("./download_GTEX/gene_tpm_2017-06-05_v8_", a, ".gct.gz", sep = "")
  dat.gct <- read.delim(file = b, skip = 2)
  d0 <- dat.gct %>%
    filter(Description %in% topgene)
  t <- rep(a, length(topgene))
  d <- cbind(t, d0)

  write.table(d, file = paste("./topgene/raw/", a, ".txt", sep = ""))
}
```

# Organize data to plot

```{r}
topgene <- c(
  "CHRNA4", "CSTB", "DEPDC5", "EPM2A", "KCNQ2",
  "KCNT1", "SCN1A", "SCN2A", "SCN8A", "KCNA2"
)
for (j in 1:length(topgene)) {
  g <- topgene[j]
  gene <- data.frame()
  for (i in 1:nrow(tissue)) {
    a <- tissue[, 2][i]
    b <- paste("./topgene/raw/", a, ".txt", sep = "")
    c <- read.table(b, sep = " ")
    d0 <- subset(c, Description == g)
    d0 <- d0[, -c(2:3)]
    d1 <- d0 %>%
      tidyr::gather(3:ncol(d0), key = ID, value = exp)
    gene <- rbind(gene, d1)
  }
  write.table(gene, file = paste("./topgene/process/", g, ".txt", sep = ""))
}
```

# Plot

```{r}
topgene <- c(
  "CHRNA4", "CSTB", "DEPDC5", "EPM2A", "KCNQ2",
  "KCNT1", "SCN1A", "SCN2A", "SCN8A", "KCNA2"
)

for (j in 1:length(topgene)) {
  g <- topgene[j]

  path <- paste("./topgene/process/", g, ".txt", sep = "")
  data <- read.table(path)

  a <- c(
    "brain_cerebellar_hemisphere", "brain_cerebellum", "brain_cortex",
    "brain_frontal_cortex_ba9", "brain_anterior_cingulate_cortex_ba24",
    "brain_caudate_basal_ganglia", "brain_putamen_basal_ganglia", "brain_nucleus_accumbens_basal_ganglia", "brain_amygdala", "brain_hippocampus",
    "brain_hypothalamus", "brain_spinal_cord_cervical_c-1",
    "brain_substantia_nigra", "pituitary", "testis",
    "esophagus_gastroesophageal_junction", "esophagus_muscularis",
    "colon_transverse", "small_intestine_terminal_ileum",
    "colon_sigmoid", "stomach", "artery_aorta", "artery_tibial",
    "artery_coronary", "adipose_subcutaneous", "adipose_visceral_omentum",
    "breast_mammary_tissue", "nerve_tibial", "cervix_ectocervix",
    "cervix_endocervix", "uterus", "bladder", "vagina", "prostate",
    "fallopian_tube", "ovary", "lung", "thyroid",
    "skin_not_sun_exposed_suprapubic", "skin_sun_exposed_lower_leg",
    "esophagus_mucosa", "minor_salivary_gland", "adrenal_gland",
    "kidney_cortex", "kidney_medulla", "spleen", "heart_atrial_appendage",
    "heart_left_ventricle", "muscle_skeletal", "liver", "pancreas",
    "whole_blood"
  )

  data <- data %>% filter(t %in% a)

  data$t <- factor(data$t, levels = a)

  f <- ggplot(data, aes(x = t, y = exp)) +
    geom_boxplot() +
    theme_bw() +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      axis.line = element_line(colour = "black", size = 0.5),
      axis.ticks = element_line(colour = "black", size = 0.5),
      axis.title.x = element_blank(),
      legend.position = "none",
      axis.text.y = element_text(size = 8)
    ) + 
    ylab("Expression level") +
    ggtitle(g)

  path <- paste("./plot/", g, ".png", sep = "")
  ggsave(f, file = path, width = 6, height = 2.5)
}
```

# Statistical analysis by class

```{r}
rm(list = ls())

data <- read.csv(file = "./epilepsygenes/epilepsy_meanExp_sum.csv")
data <- na.omit(data)
data <- data[, -c(1, 25:26)]

data_l <- data %>%
  gather(3:ncol(data), key = "tissue", value = "expression")

CNS <- c(
  "brain_cerebellar_hemisphere", "brain_cerebellum", "brain_cortex",
  "brain_frontal_cortex_ba9", "brain_anterior_cingulate_cortex_ba24",
  "brain_caudate_basal_ganglia", "brain_putamen_basal_ganglia", "brain_nucleus_accumbens_basal_ganglia", "brain_amygdala", "brain_hippocampus",
  "brain_hypothalamus", "brain_spinal_cord_cervical_c.1",
  "brain_substantia_nigra", "pituitary"
)

nonCNS <- c(
  "testis", "esophagus_gastroesophageal_junction", "esophagus_muscularis",
  "colon_transverse", "small_intestine_terminal_ileum", "colon_sigmoid",
  "stomach", "artery_aorta", "artery_tibial", "artery_coronary",
  "adipose_subcutaneous", "adipose_visceral_omentum", "breast_mammary_tissue",
  "nerve_tibial", "cervix_ectocervix", "cervix_endocervix", "uterus", "bladder",
  "vagina", "prostate", "fallopian_tube", "ovary", "lung", "thyroid",
  "skin_not_sun_exposed_suprapubic", "skin_sun_exposed_lower_leg",
  "esophagus_mucosa", "minor_salivary_gland", "adrenal_gland", "kidney_cortex",
  "kidney_medulla", "spleen", "heart_atrial_appendage", "heart_left_ventricle",
  "muscle_skeletal", "liver", "pancreas", "whole_blood"
)

for (i in 1:nrow(data_l)) {
  if (data_l$tissue[i] %in% CNS) {
    data_l$class[i] <- "CNS"
  } else if (data_l$tissue[i] %in% nonCNS) {
    data_l$class[i] <- "Non-CNS"
  } else {
    data_l$class[i] <- "NA"
  }
}

# Plot by class
ave <- data_l %>%
  group_by(Group, class) %>%
  summarise(
    group_mean = mean(expression),
    total = n(),
    sd = sd(expression),
    sem = sd / sqrt(total)
  )

ave$class <- factor(ave$class, levels = c("CNS", "Non-CNS"))

f <- ggplot(data = ave, aes(x = class, y = group_mean, group = Group)) +
  geom_errorbar(
    aes(ymin = group_mean - sem, ymax = group_mean + sem),
    position = position_dodge(), color = "gray"
  ) +
  geom_bar(stat = "identity", position = position_dodge(), aes(fill = Group)) +
  scale_color_manual(values = c("#ffb60f", "#c13832", "#673bb8")) +
  scale_fill_manual(values = c("#ffb60f", "#c13832", "#673bb8")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    axis.line = element_line(colour = "black", size = 0.5),
    axis.ticks = element_line(colour = "black", size = 0.5),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 9)
  ) + 
  ylab("Expression level")

print(f)
ggsave(file = "./plot/class_mean.png", width = 3, height = 5)

# CNS
CNS <- data_l %>% filter(class == "CNS")

# Average to gene
ave <- CNS %>%
  group_by(Gene, Group) %>%
  summarise(
    group_mean = mean(expression),
    total = n(),
    sd = sd(expression),
    sem = sd / sqrt(total)
  )

pairwise.wilcox.test(ave$group_mean, ave$Group, p.adjust.method = "BH")

# Non-CNS
nonCNS <- data_l %>% filter(class == "Non-CNS")

# average to gene
ave <- nonCNS %>%
  group_by(Gene, Group) %>%
  summarise(
    group_mean = mean(expression),
    total = n(),
    sd = sd(expression),
    sem = sd / sqrt(total)
  )

pairwise.wilcox.test(ave$group_mean, ave$Group, p.adjust.method = "BH")

# Plot by group
ave <- data_l %>%
  group_by(Group, class) %>%
  summarise(
    group_mean = mean(expression),
    total = n(),
    sd = sd(expression),
    sem = sd / sqrt(total)
  )

ave$Group <- factor(ave$Group, levels = c("CEG", "DEEG", "SRG"))

f <- ggplot(data = ave, aes(x = Group, y = group_mean, group = class)) +
  geom_errorbar(
    aes(ymin = group_mean - sem, ymax = group_mean + sem),
    position = position_dodge(), color = "gray"
  ) +
  geom_bar(
    stat = "identity",
    position = position_dodge(),
    aes(color = class, fill = Group, size = 0.5)
  ) +
  scale_color_manual(values = c("#018571", "#a6611a")) +
  scale_fill_manual(values = c("#ffb60f", "#c13832", "#673bb8")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    axis.line = element_line(colour = "black", size = 0.5),
    axis.ticks = element_line(colour = "black", size = 0.5),
    axis.title.x = element_blank(),
    legend.position = "none",
    axis.text = element_text(size = 8),
    axis.title = element_text(size = 9)
  ) +
  ylab("Expression level")

print(f)
ggsave(file = "./plot/Group_mean.png", width = 3, height = 5)

# Testing by group

# CEG
CEG <- data_l %>% filter(Group == "CEG")

# Average to gene
ave <- CEG %>%
  group_by(Gene, class) %>%
  summarise(
    group_mean = mean(expression),
    total = n(),
    sd = sd(expression),
    sem = sd / sqrt(total)
  )

pairwise.wilcox.test(ave$group_mean, ave$class)

# DEEG
DEEG <- data_l %>% filter(Group == "DEEG")

# Average to gene
ave <- DEEG %>%
  group_by(Gene, class) %>%
  summarise(
    group_mean = mean(expression),
    total = n(),
    sd = sd(expression),
    sem = sd / sqrt(total)
  )

pairwise.wilcox.test(ave$group_mean, ave$class)

# SRG
SRG <- data_l %>% filter(Group == "SRG")

# Average to gene
ave <- SRG %>%
  group_by(Gene, class) %>%
  summarise(
    group_mean = mean(expression),
    total = n(), sd = sd(expression),
    sem = sd / sqrt(total)
  )

pairwise.wilcox.test(ave$group_mean, ave$class)
```

# Statistical analysis by tissue

```{r}
rm(list = ls())

data <- read.csv(file = "./epilepsygenes/epilepsy_meanExp_sum.csv")
data <- na.omit(data)
data <- data[, -c(1, 25:26)]

data_l <- data %>% gather(3:ncol(data), key = "tissue", value = "expression")

tissue <- unique(data_l$tissue)

for (i in 1:nrow(data_l)) {
  t <- tissue[i]
  data1 <- data_l %>%
    filter(tissue == t)
  s <- pairwise.wilcox.test(data1$expression, data1$Group, p.adjust.method = "BH")
  p0 <- s$p.value
  write.csv(p0, file = paste("./pvalue/", t, ".csv", sep = ""))
}
```
